### ISSUE TITLE
FunctionalCPD argument contains self

### ISSUE THREAD

**Issue by jpfairbanks** [sisl/BayesNets.jl#59]
As a follow up to #58, I see that the FunctionalCPD function is passed the assignment of all the values, including value of the current node.

```julia
type FunctionalCPD{D} <: CPD{D}
    target::NodeName
    parents::NodeNames
    accessor::Function # calling this gives you the distribution from the assignment

    FunctionalCPD(target::NodeName, accessor::Function) = new(target, NodeName[], accessor)
    FunctionalCPD(target::NodeName, parents::NodeNames, accessor::Function) = new(target, parents, accessor)
end

name(cpd::FunctionalCPD) = cpd.target
parents(cpd::FunctionalCPD) = cpd.parents
@define_call FunctionalCPD
@compat (cpd::FunctionalCPD)(a::Assignment) = cpd.accessor(a)
```

But the LinearGaussianCPD iterates over just the parents of the node.

```julia
type LinearGaussianCPD <: CPD{Normal}
    target::NodeName
    parents::NodeNames

	a::Vector{Float64}
	b::Float64
    σ::Float64
end
LinearGaussianCPD(target::NodeName, μ::Float64, σ::Float64) = LinearGaussianCPD(target, NodeName[], Float64[], μ, σ)

name(cpd::LinearGaussianCPD) = cpd.target
parents(cpd::LinearGaussianCPD) = cpd.parents
nparams(cpd::LinearGaussianCPD) = length(cpd.a) + 2
@define_call LinearGaussianCPD
@compat function (cpd::LinearGaussianCPD)(a::Assignment)

    # compute A⋅v + b
    μ = cpd.b
    for (i, p) in enumerate(cpd.parents)
        μ += a[p]*cpd.a[i]
    end

    Normal(μ, cpd.σ)
end
```

Should the FunctionalCPD have a way to distinguish its parents from the rest of the variables in the assignment?

I am trying to have a lot of nodes that all use the same function for the FunctionalCPD, but that function depends on the parents and not the current value.

I can't encode the names of the parents into the Function passed to the FunctionalCPD, because that would require 1 Function object per node.

I guess the officially recommended way to do this is implement my own type <: CPD{D}.

**Comment by tawheeler** [comment]
Right, you can always define your own CPDs.

```
type FunctionalCPDWithParents{D} <: CPD{D}
    target::NodeName
    parents::NodeNames
    accessor::Function # calling this gives you the distribution from the assignment

    FunctionalCPDWithParents(target::NodeName, accessor::Function) = new(target, NodeName[], accessor)
    FunctionalCPDWithParents(target::NodeName, parents::NodeNames, accessor::Function) = new(target, parents, accessor)
end

name(cpd::FunctionalCPDWithParents) = cpd.target
parents(cpd::FunctionalCPDWithParents) = cpd.parents
@define_call FunctionalCPDWithParents
@compat (cpd::FunctionalCPDWithParents)(a::Assignment) = cpd.accessor(a, parents(cpd))

**Comment by jpfairbanks** [comment]
After writing this, but before reading your response, I did the same thing, but I called mine ParentFunctionalCPD. 👍  Is this something you would like to add the the set of available CPDs? I have a test case for it.

**Comment by tawheeler** [comment]
Sure! It can help illustrate that one can write their own adjustment CPDs to fit their needs. Could you submit a PR?

**Comment by jpfairbanks** [comment]
#62 adds it to the library, do you want it in the docs too?

**Comment by tawheeler** [comment]
Yes please! You can probably just add it to the table.
Thank you!

**Comment by tawheeler** [comment]
Fixed with b42506663cafe47146e4d9dfb2911110f6b84054
